server/src/test/java/org/elasticsearch/health/node/ShardsCapacityHealthIndicatorServiceTests.java

```diff
@@ -0,0 +1,407 @@
+                        "frozen",
+                        Map.of("max_shards_in_cluster", maxShardsPerNodeFrozen)
+                    )
+                )
+            );
+        }
+        {
+            // Only frozen_nodes does not have enough space
+            int maxShardsPerNode = randomValidMaxShards();
+            var clusterService = createClusterService(maxShardsPerNode, 25, createIndexInFrozenNode(11));
+            var indicatorResult = new ShardsCapacityHealthIndicatorService(clusterService).calculate(true, HealthInfo.EMPTY_HEALTH_INFO);
+
+            assertEquals(indicatorResult.status(), RED);
+            assertEquals(
+                indicatorResult.symptom(),
+                "Cluster is close to reaching the configured maximum number of shards for frozen nodes."
+            );
+            assertThat(indicatorResult.impacts(), equalTo(RED_INDICATOR_IMPACTS));
+            assertThat(indicatorResult.diagnosisList(), hasSize(1));
+            assertThat(indicatorResult.diagnosisList().get(0), equalTo(SHARDS_MAX_CAPACITY_REACHED_FROZEN_NODES));
+            assertThat(
+                xContentToMap(indicatorResult.details()),
+                is(
+                    Map.of(
+                        "data",
+                        Map.of("max_shards_in_cluster", maxShardsPerNode),
+                        "frozen",
+                        Map.of("max_shards_in_cluster", 25, "current_used_shards", 22)
+                    )
+                )
+            );
+        }
+        {
+            // Both data and frozen nodes does not have enough space
+            var clusterService = createClusterService(25, 25, createIndexInDataNode(11), createIndexInFrozenNode(11));
+            var indicatorResult = new ShardsCapacityHealthIndicatorService(clusterService).calculate(true, HealthInfo.EMPTY_HEALTH_INFO);
+
+            assertEquals(indicatorResult.status(), RED);
+            assertEquals(
+                indicatorResult.symptom(),
+                "Cluster is close to reaching the configured maximum number of shards for data and frozen nodes."
+            );
+            assertThat(indicatorResult.impacts(), equalTo(RED_INDICATOR_IMPACTS));
+            assertThat(
+                indicatorResult.diagnosisList(),
+                is(List.of(SHARDS_MAX_CAPACITY_REACHED_DATA_NODES, SHARDS_MAX_CAPACITY_REACHED_FROZEN_NODES))
+            );
+            assertThat(
+                xContentToMap(indicatorResult.details()),
+                is(
+                    Map.of(
+                        "data",
+                        Map.of("max_shards_in_cluster", 25, "current_used_shards", 22),
+                        "frozen",
+                        Map.of("max_shards_in_cluster", 25, "current_used_shards", 22)
+                    )
+                )
+            );
+        }
+    }
+
+    public void testCalculateMethods() {
+        var mockedState = mock(ClusterState.class);
```

Comment:
Please replace this with a real ClusterState. This could be `ClusterState.EMPTY_STATE`